### ISSUE TITLE
Problem with expat building exiv2 0.27.5 on Windows 10 with Visual Studio 2019; 

### ISSUE THREAD

**Issue by norbertwg** [Exiv2/exiv2#2012]
First of all: I know little about conan and cmake, but using the instructions, I had no problems to build the versions 0.27.0 up to 0.27.4. 

Now I tried to build exiv2 0.27.5 on Windows 10 with Visual Studio 2019 - and failed. 
Following the instructions from README-CONAN.md I entered
>mkdir build
>cd build
>conan install .. --build missing --profile msvc2019Release64

This gave me following error:
ERROR: Failed requirement 'zlib/1.2.11@conan/stable' from 'conanfile.py'
ERROR: Unable to find 'zlib/1.2.11@conan/stable' in remotes

Using conan search I found 'zlib/1.2.11', so I changed 
self.requires('zlib/1.2.11@conan/stable')
to
self.requires('zlib/1.2.11')

Rerunning conan install, I got similar errors, so I changed:
'gtest/1.8.1@bincrafters/stable' to 'gtest/1.8.1'
'Expat/2.2.6@pix4d/stable' to 'expat/2.2.7'

Then conan install finished without errors and I started
>cmake .. -G "Visual Studio 16 2019"

Here I get following error:
CMake Error at C:/Program Files/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find EXPAT (missing: EXPAT_LIBRARY) (found version "2.2.7")
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.20/Modules/FindEXPAT.cmake:65 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  cmake/findDependencies.cmake:45 (find_package)
  CMakeLists.txt:67 (include)
  
As I had to increase the version number for expat in conanfile.py to get run conan install successfully, I thought, I now miss a corresponding update for cmake, but did not find where to do it.

**Comment by clanmills** [comment]
@norbertwg I'm a little surprised that nobody's discussed this with you.

For sure, this is conan trouble.  The build server uses conan to build the libraries that are published on exiv2.org.  I am confident that the documentation is correct.  The build server uses --build missing and normally finds the dependencies in the conan cache.

I've attempted a fresh build with `--build Force` and encountered similar issues to you.

```cmd
C:\Users\rmills\build\v0.27.5\build>conan install .. --build Force --profile msvc2019Release64
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=16
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

zlib/1.2.11: Not found in local cache, looking in remotes...
zlib/1.2.11: Trying with 'conan-center'...
ERROR: Failed requirement 'zlib/1.2.11' from 'conanfile.py'
ERROR: HTTPSConnectionPool(host='conan.bintray.com', port=443): Max retries exceeded with url: /v1/ping (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:1122)')))

Unable to connect to conan-center=https://conan.bintray.com
1. Make sure the remote is reachable or,
2. Disable it by using conan remote disable,
Then try again.

C:\Users\rmills\build\v0.27.5\build>
```

Can you try asking on the conan site about this.  It looks like a certificate issue on conan.bintray.com

**Comment by clanmills** [comment]
https://blog.conan.io/2021/09/03/conancenter-declare-bintray-obsolete.html

I'am with `conan remote`.  In 4 years of using conan to build with Visual Studio, this is the first time I'm had to think about `conan remote`.  I'll give you an update when I get this working.

**Comment by clanmills** [comment]
I think you'll have to modify the conan remote list.  I'm using `http://center.....` to evade the certificate issue when using `https://center....`

```cmd
C:\Users\rmills\build\v0.27.5\build>conan search zlib
Existing package recipes:

zlib/1.2.11@conan/stable

C:\Users\rmills\build\v0.27.5\build>conan search gtest
Existing package recipes:

gtest/1.8.1@bincrafters/stable

C:\Users\rmills\build\v0.27.5\build>conan search expat
Existing package recipes:

Expat/2.2.6@pix4d/stable

C:\Users\rmills\build\v0.27.5\build>
```

Here's the patch for conanfile.py:

```
C:\Users\rmills\build\v0.27.5>conan remote list
conan-center: http://center.conan.io [Verify SSL: True]

C:\Users\rmills\build\v0.27.5>git diff conanfile.py
diff --git a/conanfile.py b/conanfile.py
index 6946e3df..60dbc8d3 100644
--- a/conanfile.py
+++ b/conanfile.py
@@ -21,7 +21,7 @@ class Exiv2Conan(ConanFile):
         self.options['gtest'].shared = False

     def requirements(self):
-        self.requires('zlib/1.2.11')
+        self.requires('zlib/1.2.11@conan/stable')

         if self.options.webready:
             self.requires('libcurl/7.79.0')
@@ -30,14 +30,14 @@ class Exiv2Conan(ConanFile):
             self.requires('libiconv/1.16')

         if self.options.unitTests:
-            self.requires('gtest/1.10.0')
+            self.requires('gtest/1.8.1@bincrafters/stable')
             if self.settings.build_type == "Debug":
                 self.options['gtest'].debug_postfix = ''

         if self.options.xmp:
             self.requires('XmpSdk/2016.7@piponazo/stable') # from conan-piponazo
         else:
-            self.requires('expat/2.4.1')
+            self.requires('Expat/2.2.6@pix4d/stable')

     def imports(self):
         self.copy('*.dll', dst='bin', src='bin')

C:\Users\rmills\build\v0.27.5>
```

I've kicked off the build with the command and run into trouble.

```cmd
C:\Users\rmills\build\v0.27.5\build>cmake         .. -G "Visual Studio 16 2019" -A x64
-- The CXX compiler identification is MSVC 19.29.30136.0
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Python3: C:/Python39/python3.exe (found version "3.9.0") found components: Interpreter
-- Conan: Using autogenerated Findzlib.cmake
-- Found zlib: 1.2.11 (found version "1.2.11")
-- Library zlib found C:/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/d057732059ea44a47760900cb5e4855d2bea8714/lib/zlib.lib
-- Found: C:/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/d057732059ea44a47760900cb5e4855d2bea8714/lib/zlib.lib
-- Conan: Using autogenerated FindExpat.cmake
-- Found Expat: 2.2.6 (found version "2.2.6")
-- Library expat found C:/Users/rmills/.conan/data/Expat/2.2.6/pix4d/stable/package/eb4102e772b530edfbc4b0e1be0c1f346423e000/lib/expat.lib
-- Found: C:/Users/rmills/.conan/data/Expat/2.2.6/pix4d/stable/package/eb4102e772b530edfbc4b0e1be0c1f346423e000/lib/expat.lib
-- Looking for mmap
-- Looking for mmap - not found
-- Looking for munmap
-- Looking for munmap - not found
-- Looking for strerror_r
-- Looking for strerror_r - not found
-- Performing Test EXV_STRERROR_R_CHAR_P
-- Performing Test EXV_STRERROR_R_CHAR_P - Failed
-- Looking for C++ include libproc.h
-- Looking for C++ include libproc.h - not found
-- Looking for C++ include unistd.h
-- Looking for C++ include unistd.h - not found
-- Looking for C++ include sys/mman.h
-- Looking for C++ include sys/mman.h - not found
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Failed
-- Performing Test COMPILER_HAS_DEPRECATED
-- Performing Test COMPILER_HAS_DEPRECATED - Success
-- Install prefix:    C:/Program Files (x86)/exiv2
-- ------------------------------------------------------------------
-- CMake Generator:   Visual Studio 16 2019
-- CMAKE_BUILD_TYPE:  Release
-- Compiler info: MSVC (C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe) ; version: 19.29.30136.0
-- CMAKE_CXX_STANDARD:11
--  --- Compiler flags ---
-- General:           /DWIN32 /D_WINDOWS /W3 /GR /EHsc
         /MP
         /Zc:__cplusplus
-- Extra:
-- Debug:             /MDd /Zi /Ob0 /Ox /Zo
-- Release:           /MD /O2  /DNDEBUG
-- RelWithDebInfo:    /MD /Zi /O2  /DNDEBUG
-- MinSizeRel:        /MD /O1  /DNDEBUG
--  --- Linker flags ---
-- General:           /machine:x64
-- Debug:             /debug /INCREMENTAL
-- Release:           /INCREMENTAL:NO
-- RelWithDebInfo:    /debug /INCREMENTAL
-- MinSizeRel:        /INCREMENTAL:NO
--
-- Compiler Options
-- Warnings as errors:                 NO
-- Use extra compiler warning flags:   NO
--
-- ------------------------------------------------------------------
-- Building shared library:            YES
-- Building PNG support:               YES
-- XMP metadata support:               YES
-- Building BMFF support:              YES
-- Native language support:            NO
-- Conversion of Windows XP tags:      YES
-- Nikon lens database:                YES
-- Building webready support:          NO
-- Dynamic runtime override:           YES
-- Unicode paths (wstring):            NO
-- Building exiv2 command:             YES
-- Building samples:                   YES
-- Building unit tests:                NO
-- Building fuzz tests:                NO
-- Building doc:                       NO
-- Building with coverage flags:       NO
-- Using ccache:                       NO
-- Configuring done
CMake Error at xmpsdk/CMakeLists.txt:1 (add_library):
  Target "exiv2-xmp" links to target "EXPAT::EXPAT" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at xmpsdk/CMakeLists.txt:1 (add_library):
  Target "exiv2-xmp" links to target "EXPAT::EXPAT" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
```

@piponazo Can you help?

**Comment by norbertwg** [comment]
@clanmills thanks for your feedback
I took the sources from the release asset https://github.com/Exiv2/exiv2/archive/refs/tags/v0.27.5.tar.gz. There the versions for zlib etc. are as you listed in your patch. But with these versions I had the problem, searched the web and found a hint to use the command 
   conan search zlib -r all
The output:
Remote 'conan-center':
   zlib/1.2.8
   zlib/1.2.11
This was the reason why I changed the respective entries in conanfile.py.
As "conan search expat -r all" did not list expat 2.2.6, I changed to 2.2.7 - assuming, that 2.2.6 no longer is available on conan server.

When I now enter "conan search zlib" I get:
   There are no packages matching the 'zlib' pattern

My theory: I have the problem, because I recently cleaned up my system and cleared the conan cache. And now the versions referenced in conanfile.py are no longer available on http://center.conan.io, so I cannot get them. Does this make sense?

After this I had a look into the latest version of conanfile.py in main branch and found there several changes compared to the version in the 0.27.5.tar.gz. So I tried this one. "conan install" completed without errors, but 
   cmake .. -G "Visual Studio 16 2019"
does still not work, now gets an error related to ZLIB:
CMake Error at C:/Program Files/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.20/Modules/FindZLIB.cmake:120 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  cmake/findDependencies.cmake:19 (find_package)
  CMakeLists.txt:67 (include)

**Comment by clanmills** [comment]
I've successfully build this by:

1.  Upgraded conan to 1.43.0 (as documented in README-CONAN.md)
2. Deleted the conan cache (%USERPROFILE%\\.conan\\data)
3. Reverted the code changes to conanfile.py
4. Set conan-center to https://center.conan.io

Do the conan/cmake dance:

```cmd
C:\Users\rmills\build\v0.27.5\build>conan --version
Conan version 1.43.0

C:\Users\rmills\build\v0.27.5\build>cmake --version
cmake version 3.16.5

CMake suite maintained and supported by Kitware (kitware.com/cmake).

C:\Users\rmills\build\v0.27.5\build>conan remote list
conan-center: https://center.conan.io [Verify SSL: True]

C:\Users\rmills\build\v0.27.5\build> conan install .. --build missing --profile msvc2019Release64
... brrrrrrr ...

C:\Users\rmills\build\v0.27.5\build> cmake .. -G "Visual Studio 16 2019" -A x64
... brrrrrrr ...

C:\Users\rmills\build\v0.27.5\build> cmake --build .
... brrrrrrr ...

C:\Users\rmills\build\v0.27.5\build>
```

I'm going to close this because I don't want Luis (@piponazo) to spend time on this.  The code is good.  It's one of these _"the build is broken for totally unknowable reasons"_ kind of things.

I'm happy to continue to discuss this, however let's close it for now.

**Comment by norbertwg** [comment]
@clanmills thanks a lot
I am now able to build 0.27.5 using https://github.com/Exiv2/exiv2/archive/refs/tags/v0.27.5.tar.gz and replacing the included conanfile.py by the latest version in main branch. 
Initially it still failed following your latest instructions. But then I remembered an information, when I upgraded conan before starting to build 0.27.5: settings.yml could not be overwritten. At that point of time I thought, this is not done in order not to overwrite my user settings. Now I compared the file with the created .new and decided to use the new version - and it worked. Seems this was the reason for the second problem I faced (after solving the first problem by replacing conanfile.py). 

Another information: Before I read your latest post I tried to build the Exiv2 nightly prerelease, which succeeded immediately. So my problems were related to 0.27.5 only.

**Comment by clanmills** [comment]
Thanks for the update.  I also had to replace some xml magic in conan relating to certificates.  I didn't want to discuss that with you because I thought "that's something to do with my machine".

I successfully built using conanfile.py in the released product.  Can you revisit that, please?  I believe the released code works.

**Comment by kevinbackhouse** [comment]
Looks like Conan is causing similar issues in our automated builds for the 0.27-maintenance branch. See #2016.

**Comment by clanmills** [comment]
I'll investigate.  I think we perform three more steps at the start of the build:

1.  pip install conan --upgrade
2. conan remote remove conan-center ; conan remote add conan-center https://center.conan.io
3. Update ~/.conan/settings.yml and ~/.conan/cacert.pem (if settings.yml.new and cacert.pem.new exist).

Where's the code for the CI and I'll deal with this after lunch?

**Comment by norbertwg** [comment]
@clanmills : I included 0.27.5 lib in my Windows application and did tests there. Did not find any issues.

**Comment by clanmills** [comment]
Thank You for the feedback @norbertwg.  That's good to hear.

I'm having a tough week with #2017.  I've only been involved with the CI for MinGW and Cygwin and they build using the host package manager.  I have found numerous errors concerning building with conan.

When I request a review of the changes on branch fix_conan_ci (which is derived from 0.27-maintenance) it would be very helpful if you could build and provide feedback.  I hope branch fix_conan_ci will be merged into 0.27-maintenance in the next week.

**Comment by clanmills** [comment]
@norbertwg Luis has serviced conanfile.py (and the CI) on branch 0.27-maintenance.  #2018 has been merged.  Our CI builds many targets including the Windows targets Visual Studio 2015, 2017 and 2019, Cygwin/64 and MinGW/msys2, as well as lots of Linux distos and macOS.  The CI Visual Studio and Linux builds use conan to build dependencies.  The CI macOS, Cygwin and MinGW builds use the platform manager to install dependences.

So our CI has successfully built and tested several version of Visual Studio.  I have built Visual Studio 2019 both on my machine and on our build server.  I believe everything is working as documented.

It would be appreciated if you could sync and build 0.27-maintenance with Visual Studio.  Team Exiv2 has been using conan 1.43.0 and I believe you have already upgraded to that.

**Comment by norbertwg** [comment]
@clanmills I synced and built 0.27-maintenance with Visual Studio 2019, then included the lib in my application and ran some tests and found no issues. Confirmed: I upgraded to conan 1.43.0

**Comment by clanmills** [comment]
@norbertwg Thanks for the update. Pleased to hear that everything's working correctly.  That's our aim of course.
